### PR TITLE
[Spec Decode] Support SWA + DFlash for MiMo

### PR DESCRIPTION
**Pull request by benchislett**
## Purpose

Adds support for various combinations of causality and sliding-window for DFlash layers.
Enables all-causal SWA and all-non-causal SWA for DFlash, and sets up a model-implementation foundation from which to build out hybrid SWA+Full DFlash support in MRV2.

Also adds support for loading the mask embedding from a file ("mask_embedding.pt") in the model checkpoint path; since the [MiMo checkpoint](https://huggingface.co/XiaomiMiMo/MiMo-V2.5-Pro-FP4-DFlash) does not properly set it in the vocabulary embedding table. 

Limitation: Currently the model needs to be downloaded explicitly in order to properly run the DFlash:
```
vllm serve ./mimov25/ \
    --tensor-parallel-size 4 \
    --trust-remote-code \
    --gpu-memory-utilization 0.95 \
    --max-model-len auto \
    --reasoning-parser mimo \
    --tool-call-parser mimo \
    --enable-auto-tool-choice \
    --no-enable-prefix-caching \
    --generation-config vllm \
    --speculative_config '{"method":"dflash","model":"./mimov25/dflash","num_speculative_tokens":7}'
```

I think this is a limitation given the format which MiMo used to release the DFlash code. Not sure if there's a better way.

## Test Plan

Serve with:
```
vllm serve ./mimov25/ \
    --tensor-parallel-size 4 \
    --trust-remote-code \
    --gpu-memory-utilization 0.95 \
    --max-model-len auto \
    --reasoning-parser mimo \
    --tool-call-parser mimo \
    --enable-auto-tool-choice \
    --no-enable-prefix-caching \
    --generation-config vllm \
    --speculative_config '{"method":"dflash","model":"./mimov25/dflash","num_speculative_tokens":7, "attention_backend": "FLASH_ATTN"}' \
    --no-async-scheduling
```

Benchmarks:
```
vllm bench serve --dataset-name speed_bench --dataset-path . --speed-bench-dataset-subset qualitative --num-prompts -1 --save-result --save-detailed --disable-shuffle --max-concurrency 32 --temperature 1.0 --speed-bench-output-len 2048 --top-p 0.95

vllm bench serve --dataset-name speed_bench --dataset-path . --speed-bench-dataset-subset throughput_2k --num-prompts -1 --save-result --save-detailed --disable-shuffle --max-concurrency 32 --temperature 1.0 --speed-bench-output-len 1024 --top-p 0.95

vllm bench serve --dataset-name speed_bench --dataset-path . --speed-bench-dataset-subset throughput_32k --num-prompts -1 --save-result --save-detailed --disable-shuffle --max-concurrency 32 --temperature 1.0 --speed-bench-output-len 1024 --top-p 0.95
```

## Test Result

All completions seem sensible. Acceptance Rates:
- SPEED-Bench (Coding Subset): **3.73**
- SPEED-Bench (Qualitative Overall Average): **3.10**
- SPEED-Bench (2k ISL): **2.67**
- SPEED-Bench (32k ISL): **2.61**

### CODE REVIEW

File: vllm/model_executor/models/qwen3_dflash.py [hunk 1]
```diff
@@ -109,16 +180,33 @@ def __init__(
             max_position=max_position,
             rope_parameters=rope_parameters,
         )
+
+        self.attention_sink_bias = (
+            torch.nn.Parameter(torch.empty(self.num_heads), requires_grad=False)
+            if add_swa_attention_sink_bias
+            else None
+        )
```
Comment (TheEpicDolphin) [inline vllm/model_executor/models/qwen3_dflash.py:195]: It seems like this parameter is not being used anywhere. I think we need to pass it into Attention and also make sure it gets sharded among TP ranks during `load_weights()`.

File: vllm/model_executor/models/qwen3_dflash.py [hunk 2]
```diff
@@ -47,6 +52,74 @@
...
+    - Without ``layer_types``, every layer is full attention unless
+      ``dflash_config.use_swa`` is set, in which case every layer is SWA.
+    - With a ``layer_types`` carrying sliding entries, each layer follows its
+      own entry. An all-"full_attention" ``layer_types`` may be synthesized
+      when the checkpoint omits one, so ``use_swa`` still forces every layer to
+      SWA when ``layer_types`` has no sliding entries.
+
+    Causality: ``dflash_config.causal`` applies to all layers when set.
+    Otherwise full-attention layers are non-causal and SWA layers are causal;
+    uniform full/SWA models default to non-causal.
```
Comment (TheEpicDolphin) [inline vllm/model_executor/models/qwen3_dflash.py]: Sorry for the noob question, but do/will any DFlash models have causal attention layers? I thought that by nature they all used bidirectional attention in all layers?

Comment (benchislett) [reply vllm/model_executor/models/qwen3_dflash.py]: Yes. For one, `z-lab/gemma-4-31B-it-DFlash` and all the Qwen3.5 DFlash have a mix of causal-SWA and non-causal full attention
https://github.com/vllm-project/vllm/pull/40898#discussion_r3184827669

Comment (benchislett) [reply vllm/model_executor/models/qwen3_dflash.py]: And myself (and speculators team, I believe) have trained many checkpoints using causal within the query and it works as-good-or-better than non-causal DFlash, and is easier to support with kernels. Just haven't shipped any yet, but it's easy and useful to have support for R&D regardless

**Comment by benchislett** [comment]
base model slightly bugged with specdec, unrelated to DFlash.
https://github.com/vllm-project/vllm/issues/46669

**Review by TheEpicDolphin — APPROVED** [review]
LGTM

File: vllm/model_executor/models/qwen3_dflash.py [hunk 3]
```diff
@@ -109,16 +185,34 @@ def __init__(
...
+            per_layer_sliding_window=sliding_window,
             prefix=f"{prefix}.attn",
             attn_type=attn_type,
+            sinks=self.attention_sink_bias,
         )
+        if sliding_window is not None and not causal:
+            # Dedicated override for FlashAttention + non-causal SWA:
+            # we need to ensure that the sliding window is symmetric around the query
+            # (a sliding-window both forwards and backwards). This is a rare edge-case.
+            # FA3 can sometimes build the config incorrectly, so patch it here anyways.
```
Comment (LucasWilkinson) [inline vllm/model_executor/models/qwen3_dflash.py]: can you please elaborate on why the config builds incorrectly sometimes?

Comment (benchislett) [reply vllm/model_executor/models/qwen3_dflash.py]: FA3 has some weird behaviour, I think it doesn't apply the right window in fast_build=False, because of AOT scheduling? Something like that

Comment (benchislett) [reply vllm/model_executor/models/qwen3_dflash.py]: I think it was this?
https://github.com/vllm-project/vllm/blob/d490b98162422922230c8b91757c58077e24682f/vllm/v1/attention/backends/flash_attn.py#L290

Comment (benchislett) [reply vllm/model_executor/models/qwen3_dflash.py]: Admittedly I don't fully understand how FA does bidirectional sliding window. This just seemed like the safest option

Comment (benchislett) [reply vllm/model_executor/models/qwen3_dflash.py]: See also https://github.com/vllm-project/vllm/blob/d490b98162422922230c8b91757c58077e24682f/vllm/v1/attention/backends/flash_attn.py#L655

**Comment by benchislett** [comment]
oops misclick

File: vllm/model_executor/models/qwen3_dflash.py (hunk 3 above)
Comment (TheEpicDolphin) [reply vllm/model_executor/models/qwen3_dflash.py]: It seems like `FlashAttentionImpl` assumes that if `attn_type` is DECODER type, then the sliding window is shaped for causal attention, hence why we get (sliding_window - 1, 0)? Maybe the proper fix is to thread causality through to `FlashAttentionImpl.__init__` and build the sliding window accordingly?

If that's a difficult refactor, then I think this patch makes sense for now. We can follow up with the appropriate fix for the sliding window shape later?

Comment (TheEpicDolphin) [reply vllm/model_executor/models/qwen3_dflash.py]: It seems like there was an attempt to address this asymmetric window size for non-causal models here in the forward pass:
https://github.com/vllm-project/vllm/blob/a2e8ec3d52ab4e163501c8c7bee8c03ca8359a7a/vllm/v1/attention/backends/flash_attn.py#L834-L845

But it looks like that runs too late because AOT scheduling already read the `sliding_window` value before that

Comment (benchislett) [reply vllm/model_executor/models/qwen3_dflash.py]: Changed it around so that the sliding window gets built and cached on the metadata

File: vllm/v1/attention/backends/flash_attn.py [hunk 4]
```diff
@@ -822,28 +848,16 @@ def forward(
...
                 )
+                sliding_window_size = list(window) if window is not None else None
 
                 causal = attn_metadata.causal
                 is_dynamic_causal = isinstance(causal, torch.Tensor)
 
-                # For non-causal (bidirectional) attention, make the
-                # sliding window symmetric so queries attend in both
-                # directions.
-                if (
```
Comment (benchislett) [inline vllm/v1/attention/backends/flash_attn.py:859]: Note, this now happens in `build()`, and we read the (already corrected) sliding window from the metadata

File: vllm/model_executor/models/qwen3_dflash.py (hunk 2 above)
Comment (LucasWilkinson) [inline vllm/model_executor/models/qwen3_dflash.py]: nit: i find this comment a bit heavy to read, i think simple table like
```
"""Resolve ``(sliding_window, causal)`` for one DFlash draft layer.

+-----------------------------------------------+---------------------------+--------------------------+
| Config                                        | ``layer_type``            | ``causal``               |
+===============================================+===========================+==========================+
| ``layer_types`` contains sliding entries      | Use this layer's type     | SWA: ``True``            |
|                                               |                           | Full: ``False``          |
| No sliding entries + ``use_swa=True``         | SWA for every layer       | ``False``                |
| No sliding entries + ``use_swa=False``        | Full for every layer      | ``False``                |
+-----------------------------------------------+---------------------------+--------------------------+

If ``dflash_config.causal`` is set, its value overrides ``causal`` for all layers.
"""
```
could make this much easier to read

**Review by LucasWilkinson — APPROVED** [review]
LGTM, left one nit

File: vllm/model_executor/models/qwen3_dflash.py (hunk 2 above)
Comment (benchislett) [reply vllm/model_executor/models/qwen3_dflash.py]: good point, it's a bit hairy
